File: tests/bench/crosslang/go/bank.go
// Порт tests/bench/programs/BankBench.eat 1:1 — банкованная память
// pool[a/4096][a%4096], адрес зависит от loop-carried acc. Вывод
// обязан совпасть байт-в-байт с EATLang-оригиналом при том же REPEAT.
package main

import "fmt"

const REPEAT uint32 = 1

func work(seed uint32) uint32 {
	var pool [8][4096]uint32
	acc := seed
	for k := uint32(0); k < 200; k++ {
		for i := uint32(0); i < 1000; i++ {
			a := (acc + i*37) % 32768
			pool[a/4096][a%4096] = (acc + i) % 65536
			b := (a*17 + 5) % 32768
			acc = (acc + pool[b/4096][b%4096]) % 65536
		}
	}
	return acc
}

func main() {
	acc := uint32(7)
	for r := uint32(0); r < REPEAT; r++ {
		acc = work(acc)
	}
	fmt.Printf("checksum %d\n", acc)
}

File: tests/bench/crosslang/c/bank.c
/* Порт tests/bench/programs/BankBench.eat 1:1 — банкованная память
 * pool[a/4096][a%4096], адрес зависит от loop-carried acc (LLVM не
 * сворачивает). Вывод обязан совпасть байт-в-байт с EATLang-
 * оригиналом при том же REPEAT. Банк — static (128 КБ, не на стеке
 * main), сбрасывается в начале каждого work() как var pool в EATLang. */
#include <stdint.h>
#include <stdio.h>
#include <string.h>

static const uint32_t REPEAT = 1;

static uint32_t pool[8][4096];

static uint32_t work(uint32_t seed) {
    memset(pool, 0, sizeof(pool));
    uint32_t acc = seed;
    for (uint32_t k = 0; k < 200u; k++) {
        for (uint32_t i = 0; i < 1000u; i++) {
            uint32_t a = (acc + i * 37u) % 32768u;
            pool[a / 4096u][a % 4096u] = (acc + i) % 65536u;
            uint32_t b = (a * 17u + 5u) % 32768u;
            acc = (acc + pool[b / 4096u][b % 4096u]) % 65536u;
        }
    }
    return acc;
}

int main(void) {
    uint32_t acc = 7u;
    for (uint32_t r = 0; r < REPEAT; r++) {
        acc = work(acc);
    }
    printf("checksum %u\n", acc);
    return 0;
}
